dffml/util/testing/source.py

```diff
@@ -60,40 +59,78 @@ class SourceTest(abc.ABC):
-                        )
-                        self.assertEqual(repo.data.prediction.confidence, 0.42)
-                    with self.subTest(src_url=empty_src_url):
-                        repo = await sourceContext.repo(empty_src_url)
-                        self.assertFalse(repo.data.prediction.classification)
-                        self.assertFalse(repo.data.prediction.confidence)
-                    with self.subTest(both=[full_src_url, empty_src_url]):
-                        repos = {
-                            repo.src_url: repo
-                            async for repo in sourceContext.repos()
-                        }
-                        self.assertIn(full_src_url, repos)
-                        self.assertIn(empty_src_url, repos)
-                        self.assertEqual(
-                            repos[full_src_url].features(),
-                            full_repo.features(),
-                        )
-                        self.assertEqual(
-                            repos[empty_src_url].features(),
-                            empty_repo.features(),
-                        )
+        source = await self.setUpSource()
+        async with source as testSource:
+            # Open, update, and close
+            async with testSource() as sourceContext:
+                await sourceContext.update(full_repo)
+                await sourceContext.update(empty_repo)
+        async with source as testSource:
+            # Open and confirm we saved and loaded correctly
+            async with testSource() as sourceContext:
+                with self.subTest(src_url=full_src_url):
+                    repo = await sourceContext.repo(full_src_url)
+                    self.assertEqual(
+                        repo.data.prediction.classification, "feedface"
+                    )
+                    self.assertEqual(repo.data.prediction.confidence, 0.42)
+                with self.subTest(src_url=empty_src_url):
+                    repo = await sourceContext.repo(empty_src_url)
+                    self.assertFalse(repo.data.prediction.classification)
+                    self.assertFalse(repo.data.prediction.confidence)
+                with self.subTest(both=[full_src_url, empty_src_url]):
+                    repos = {
+                        repo.src_url: repo
+                        async for repo in sourceContext.repos()
+                    }
+                    self.assertIn(full_src_url, repos)
+                    self.assertIn(empty_src_url, repos)
+                    self.assertEqual(
+                        repos[full_src_url].features(), full_repo.features()
+                    )
+                    self.assertEqual(
+                        repos[empty_src_url].features(), empty_repo.features()
+                    )
+
+
+class FileSourceTest(SourceTest):
+    """
+    Test case class used to test a Source implementation. Subclass from and set
```

Comment:
I guess you've missed out "Test case class used to test a Source implementation. Subclass from ___ and set..."